### PR TITLE
Refactor wps_import

### PR DESCRIPTION
**Pull request by huard**
## Overview

This PR fixes https://github.com/Ouranosinc/raven/issues/40

Changes:

- use boltons.funcutils.FunctionBuilder to create functions dynamically instead of eval/exec statements
- the import_wps function returns a class instead of a module... I felt that simplified the execution of processes. And import_wps  is equivalent to the BirdyClient class. Any method that is not private on the class is a WPS process.
- expose username, password, headers, verify and cert so that they are passed to owslib for authentication
- Raise appropriate error when unauthorized
- convert outputs (int, float, bool, datetime) to python types instead of keeping the strings returned from owslib
- It seems that all of the data types work, except the bounding box type. I've seen in a comment that owslib doesn't support it, but I'm not sure if that's actually true...
- netCDF tries to open the link as OpenDAP, and if that doesn't work the file is downloaded
- probably some more things I'm forgetting...

### CODE REVIEW

File: birdy/native/__init__.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
+"""
+The :func:`import_wps` function *imports* on the fly a python module whose
```
Comment (huard) [inline birdy/native/__init__.py]: Update the doc to reflect that the object created is a class with methods calling processes.

File: birdy/native/client.py [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+from boltons.funcutils import FunctionBuilder
+from owslib.util import ServiceException
+from owslib.wps import WPS_DEFAULT_VERSION, WebProcessingService, SYNC
+
+from birdy.exceptions import UnauthorizedException
+from birdy.native.utils import build_doc, convert_input_value, convert_output_value, delist
+from birdy.native.converters import default_converters
+
+
+# TODO: Add credentials and tokens
```
Comment (huard) [inline birdy/native/client.py]: Update todos. Do they still make sense ? Maybe the log is unnecessary with the improved error handling.

File: birdy/native/client.py [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+                )
+
+        process = self._processes[pid]
+
+        input_defaults = OrderedDict(
+            (i.identifier, getattr(i, "defaultValue", None)) for i in process.dataInputs
+        )
+
+        cleaned_locals = "{k: v for k, v in locals().items() if k not in %s}"
+        cleaned_locals = cleaned_locals % str(["self"])
```
Comment (huard) [inline birdy/native/client.py]: This bit is mysterious to me. Please add a short explanation.

File: tests/test_native.py [hunk 4]
```diff
@@ -1,75 +1,130 @@
...
 
-    m = import_wps(url=url, processes='nap')
-    assert count_mod_func(m) == 1
+    m = import_wps(url=url, processes="nap")
+    assert count_class_methods(m) == 1
 
 
 @pytest.mark.online
-def test_netcdf():
+def test_inputs():
```
Comment (huard) [inline tests/test_native.py]: Very nice !

File: birdy/native/client.py [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+        >>> emu = BirdyClient(url='<server url>')
+        >>> emu.hello('stranger')
+        'Hello stranger'
+    """
+
+    def __init__(
+        self,
+        url,
+        processes=None,
+        convert_objects=True,
```
Comment (cehbrecht) [inline birdy/native/client.py]: I would keep "output as reference" as default.

**Comment by cehbrecht** [comment]
@davidcaron @huard I tried it locally with Emu and it worked for me :) I have left a comment ... I would keep "output as reference" as default. We don't want to download large netcdf files automatically.

I like the `BirdyClient` class :)

There is something with process identifier names. A process identifier can be any kind of string ... also with spaces and special chars. We were nice in our process definitions so far and avoided such names ;) But it could happen with other WPS service. I would skip this in this PR and handle this later with a new PR.

File: birdy/native/client.py [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+                    "You are not authorized to do a request of type: GetCapabilities"
+                )
+
+        all_processes = set([p.identifier for p in self._wps.processes])
+        if processes is None:
+            processes = all_processes
+        elif isinstance(processes, six.string_types):
+            processes = [processes]
+
+        processes = set([p.lower() for p in processes])
```
[the remark this answers is not in the document]
Comment (cehbrecht) [reply birdy/native/client.py]: I would postpone process identifier handling to a later PR. Needs to be handled both for "native" and command line client. (A process identifier could be any kind of string though we were nice so far).

**Pull request merged**